File: server/dotnet/sdk-dotnet-example/Infrastructure/SDKClients/MandateClient.cs
using Business.Domain.Payments;
using Business.DTOs.CreatePayments;
using Business.Interfaces.SDKClients;
using Infrastructure.Mappers;
using Infrastructure.Mappers.Exceptions;
using OnlinePayments.Sdk;
using OnlinePayments.Sdk.Domain;
using OnlinePayments.Sdk.Merchant;
using MandateResponse = Business.Domain.Payments.MandateResponse;

namespace Infrastructure.SDKClients;

public class MandateClient(ILogger<PaymentLinkClient> logger, IMerchantClient merchantClient) : IMandateClient
{
    public async Task<Mandate> CreateMandate(CreatePaymentRequestDto mandateRequest)
    {
        try
        {
            CreateMandateRequest requestSdk = MandateMapper.Map(mandateRequest);
            logger.LogInformation("Creating mandate start.");
            CreateMandateResponse mandateResponse = await merchantClient.Mandates.CreateMandate(requestSdk);
            logger.LogInformation("Successful mandate with unique mandate reference: {}",
                mandateResponse.Mandate.UniqueMandateReference);
            Mandate response = MandateMapper.Map(mandateResponse);

            return response;
        }
        catch (ApiException ex)
        {
            throw ExceptionMapper.Map(ex);
        }
    }

    public async Task<MandateResponse?> GetMandate(string? existingUniqueMandateReference)
    {
        MandateResponse? mandate = null;
        try
        {
            logger.LogInformation("Get mandate request.");
            GetMandateResponse? mandateResponse =
                await merchantClient.Mandates.GetMandate(existingUniqueMandateReference);
            logger.LogInformation("Mandate retrieved successfully.");
            mandate = MandateMapper.Map(mandateResponse);
            
            return mandate;
        }
        catch (ReferenceException ex)
        {
            throw ExceptionMapper.Map(ex, $"Mandate with reference: {mandate?.UniqueMandateReference} not found.");
        }
    }
}

File: server/nodejs/src/infrastructure/sdk-clients/mandate-client.ts
import { FastifyBaseLogger } from 'fastify';
import { IMandateClient } from '../interfaces/mandate-client-interface';
import { CreatePaymentRequestDto } from '../../business/dtos/payment/create-payment/create-payment-request-dto';
import { Mandate } from '../../business/domain/payments/mandate';
import { MandateMapper } from '../mappers/mandate-mapper';
import { createMerchantClient } from '../../configuration/merchant-client-configuration/merchant-client-config';
import { ErrorMapper } from '../mappers/error-mapper';

export const mandateClient = (merchantClient: ReturnType<typeof createMerchantClient>): IMandateClient => ({
    async createMandate(request: CreatePaymentRequestDto, logger: FastifyBaseLogger): Promise<Mandate> {
        const client = merchantClient.getClient(logger);
        const merchantId = merchantClient.getMerchantId(logger);
        const sdkRequest = MandateMapper.toSdkRequest(request);

        logger.info('Creating mandate start.');

        const response = await client.mandates.createMandate(merchantId, sdkRequest);

        if (!response.isSuccess) {
            logger.error(
                `Failed to create mandate - Status: ${response.status}, Error: ${JSON.stringify(response.body)}`,
            );
            throw ErrorMapper.fromSdkError(response.body, response.status);
        }

        logger.info(
            `Successful mandate with unique mandate reference: ${response.body?.mandate?.uniqueMandateReference}.`,
        );

        return MandateMapper.fromSdkResponse(response.body.mandate);
    },

    async getMandate(existingUniqueMandateReference: string, logger: FastifyBaseLogger): Promise<Mandate> {
        const client = merchantClient.getClient(logger);
        const merchantId = merchantClient.getMerchantId(logger);

        logger.info('Get mandate request.');

        const response = await client.mandates.getMandate(merchantId, existingUniqueMandateReference);

        if (!response.isSuccess) {
            logger.error(`Failed to get mandate - Status: ${response.status}`);
            throw ErrorMapper.fromSdkError(
                null,
                400,
                `Mandate with reference: ${existingUniqueMandateReference} not found.`,
            );
        }

        logger.info('Mandate retrieved successfully.');

        return MandateMapper.fromSdkResponse(response.body.mandate);
    },
});
